src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1709,175 +1703,74 @@ public static unsafe nuint WidenAsciiToUtf16(byte* pAsciiBuffer, char* pUtf16Buf
+        private static unsafe nuint WidenAsciiToUtf16_Vector128(byte* pAsciiBuffer, char* pUtf16Buffer, nuint elementCount)
         {
-            // JIT turns the below into constants
-
-            uint SizeOfVector128 = (uint)Unsafe.SizeOf<Vector128<byte>>();
-            nuint MaskOfAllBitsInVector128 = (nuint)(SizeOfVector128 - 1);
-
-            // This method is written such that control generally flows top-to-bottom, avoiding
-            // jumps as much as possible in the optimistic case of "all ASCII". If we see non-ASCII
-            // data, we jump out of the hot paths to targets at the end of the method.
-
-            Debug.Assert(Sse2.IsSupported || AdvSimd.Arm64.IsSupported);
+            Debug.Assert(Vector128.IsHardwareAccelerated);
             Debug.Assert(BitConverter.IsLittleEndian);
-            Debug.Assert(elementCount >= 2 * SizeOfVector128);
-
-            // We're going to get the best performance when we have aligned writes, so we'll take the
-            // hit of potentially unaligned reads in order to hit this sweet spot.
-
-            Vector128<byte> asciiVector;
-            Vector128<byte> utf16FirstHalfVector;
-            bool containsNonAsciiBytes;
+            Debug.Assert(elementCount >= 2 * (uint)Vector128<byte>.Count);
 
-            // First, perform an unaligned read of the first part of the input buffer.
-
-            if (Sse2.IsSupported)
-            {
-                asciiVector = Sse2.LoadVector128(pAsciiBuffer); // unaligned load
-                containsNonAsciiBytes = (uint)Sse2.MoveMask(asciiVector) != 0;
-            }
-            else if (AdvSimd.Arm64.IsSupported)
-            {
-                asciiVector = AdvSimd.LoadVector128(pAsciiBuffer);
-                containsNonAsciiBytes = ContainsNonAsciiByte(asciiVector);
-            }
-            else
-            {
-                throw new PlatformNotSupportedException();
-            }
+            nuint currentOffset = 0;
+            ushort* pCurrentWriteAddress = (ushort*)pUtf16Buffer;
 
-            // If there's non-ASCII data in the first 8 elements of the vector, there's nothing we can do.
+            // Calculating the destination address outside the loop results in significant
+            // perf wins vs. relying on the JIT to fold memory addressing logic into the
+            // write instructions. See: https://github.com/dotnet/runtime/issues/33002
+            nuint finalOffsetWhereCanRunLoop = elementCount - (uint)Vector128<byte>.Count;
 
-            if (containsNonAsciiBytes)
+            do
             {
-                return 0;
-            }
+                Vector128<byte> asciiVector = Vector128.Load(pAsciiBuffer + currentOffset);
 
-            // Then perform an unaligned write of the first part of the input buffer.
+                if (asciiVector.ExtractMostSignificantBits() != 0)
```

Comment:
`if (asciiVector != Vector128<byte>.Zero)` is faster

Reply:
@EgorBo I am not sure if I understand. `ExtractMostSignificantBits` allows here for checking whether any of the bytes is > 127 (not an ascii char).

Reply:
Oops, right, it's a muscle memory for cases where this pattern is used after comparisons